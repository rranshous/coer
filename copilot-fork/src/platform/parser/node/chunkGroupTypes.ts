/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/

import { TreeSitterChunkHeaderInfo, TreeSitterOffsetRange } from './nodes';

/**
 * For workspace semantic chunking
 */

export interface GenericDetail extends DetailBlock {
	comments: TreeSitterChunkHeaderInfo[];
}
export interface PythonDetail extends DetailBlock {
	docstring?: TreeSitterChunkHeaderInfo;
	decorator?: TreeSitterChunkHeaderInfo;
}

export interface BlockNameDetail extends DetailBlock {
	name: string | undefined;
}

export interface DetailBlock {
	body: TreeSitterChunkHeaderInfo;
}

export interface MatchGroup<DetailType extends DetailBlock> {
	mainBlock: TreeSitterChunkHeaderInfo;
	detailBlocks: DetailType;
}

export interface QueryMatchNode<DetailType extends DetailBlock> {
	info: MatchGroup<DetailType>;
	children: QueryMatchNode<DetailType>[];
}

/**
 * This tree is like the `Parser.SyntaxTree` that we have when we use TreeSitter to parse,
 * but it only has the `MatchGroups` that get passed into the constructor.
 *
 * It sorts the `MatchGroups` into its own hierarchy.
 *
 * This assumes that the constructor's `blockInfos` are a part of the same syntax tree.
 */
export class QueryMatchTree<DetailType extends DetailBlock> {

	public roots: QueryMatchNode<DetailType>[] = [];

	/**
	 * @remark mutates the passed `groups`
	 */
	constructor(groups: MatchGroup<DetailType>[], public readonly syntaxTreeRoot: TreeSitterChunkHeaderInfo) {
		this.formTree(groups);
	}
	/**
	 * This assumes that overlapping blocks imply that one fully overlaps another.
	 * Runs with the same assumptions as `removeOverlapping`.
	 * @param groups to use as node content in the tree
	 */
	private formTree(groups: MatchGroup<DetailType>[]) {
		groups
			.sort((a, b) => a.mainBlock.startIndex - b.mainBlock.startIndex || a.mainBlock.endIndex - b.mainBlock.endIndex);

		const recentParentStack: QueryMatchNode<DetailType>[] = [];

		const peekParent = () => {
			return recentParentStack[recentParentStack.length - 1];
		};

		const hasEqualRange = (a: MatchGroup<DetailType>, b: MatchGroup<DetailType>) => {
			return (a.mainBlock.startIndex === b.mainBlock.startIndex &&
				a.mainBlock.endIndex === b.mainBlock.endIndex);
		};

		for (const group of groups) {
			const matchNode: QueryMatchNode<DetailType> = {
				info: group,
				children: []
			};
			let currParent = peekParent();

			if (!currParent) {
				this.roots.push(matchNode);
				recentParentStack.push(matchNode);
				continue;
			}

			if (hasEqualRange(currParent.info, group)) {
				// any duplicate nodes will always be one after another
				continue;
			}

			while (currParent && !TreeSitterOffsetRange.doesContain(currParent.info.mainBlock, group.mainBlock)) {
				recentParentStack.pop();
				currParent = peekParent();
			}



			if (currParent) {
				currParent.children.push(matchNode);
			} else {
				this.roots.push(matchNode);
			}

			recentParentStack.push(matchNode);
		}
	}

}

/**
 * A tree of block names that would appear in semantic chunks.
 */

export interface BlockNameNode {
	name: string;
	children: BlockNameNode[];
}