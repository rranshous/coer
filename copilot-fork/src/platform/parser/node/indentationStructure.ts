/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/

import type * as vscode from 'vscode';
import { CharCode } from '../../../util/vs/base/common/charCode';
import { BugIndicatingError } from '../../../util/vs/base/common/errors';
import { Position } from '../../../vscodeTypes';
import { AbstractDocument } from '../../editing/common/abstractText';
import { OverlayNode } from './nodes';

export function getStructureUsingIndentation(
	document: AbstractDocument,
	languageId: string,
	formattingOptions: vscode.FormattingOptions | undefined
): OverlayNode {
	const lines = document.getText().split(/\r\n|\r|\n/g);
	const opts = formattingOptions || { tabSize: 4 };
	const simpleModel = {
		getLineCount: () => lines.length,
		getLineContent: (lineNumber: number) => lines[lineNumber - 1],
		getOptions: () => opts
	} satisfies ISimpleTextModel;

	try {
		const regions = generateFoldingRegions(simpleModel, languageId);
		const [foldingRanges,] = createFoldingRangeTree(document, regions, undefined);
		foldingRanges.adjust(document, isOffSide(languageId));
		return foldingRanges.toOverlayNode(document, true);
	} catch (err) {
		const foldingRanges = new FoldingRangeNode(1, document.getLineCount(), []);
		return foldingRanges.toOverlayNode(document, true);
	}
}

function createFoldingRangeTree(doc: AbstractDocument, regions: FoldingRegions, regionIndex: number | undefined): [FoldingRangeNode, number] {
	if (typeof regionIndex !== 'undefined' && regionIndex >= regions.length) {
		throw new Error(`Invalid region index ${regionIndex}`);
	}

	const regionStartLineNumber = (typeof regionIndex === 'undefined' ? 1 : regions.getStartLineNumber(regionIndex));
	const regionEndLineNumber = (typeof regionIndex === 'undefined' ? doc.getLineCount() : regions.getEndLineNumber(regionIndex));
	const children: FoldingRangeNode[] = [];
	let childNode: FoldingRangeNode | null = null;

	regionIndex = (typeof regionIndex === 'undefined' ? 0 : regionIndex + 1);
	while (regionIndex < regions.length) {
		const startLineNumber = regions.getStartLineNumber(regionIndex);
		const endLineNumber = regions.getEndLineNumber(regionIndex);

		if (startLineNumber > regionEndLineNumber || endLineNumber > regionEndLineNumber) {
			// We are done with children of this region
			break;
		}

		const prevChildNode = childNode;
		[childNode, regionIndex] = createFoldingRangeTree(doc, regions, regionIndex);
		if (prevChildNode && childNode.startLineNumber <= prevChildNode.endLineNumber) {
			throw new BugIndicatingError('Invalid Folding Ranges: overlapping children');
		}
		if (childNode.startLineNumber < regionStartLineNumber) {
			throw new BugIndicatingError('Invalid Folding Ranges: child starts before parent');
		}
		children.push(childNode);
	}

	return [
		new FoldingRangeNode(regionStartLineNumber, regionEndLineNumber, children),
		regionIndex,
	];
}

class FoldingRangeNode {

	constructor(
		public startLineNumber: number,
		public endLineNumber: number,
		readonly children: FoldingRangeNode[]
	) {
		if (startLineNumber > endLineNumber) {
			throw new BugIndicatingError('Invalid Folding Ranges: startLineNumber > endLineNumber');
		}
	}

	public adjust(document: AbstractDocument, isOffside: boolean): void {
		if (isOffside) {
			this._adjustOffside();
		} else {
			this._adjustRegular(document, document.getLineCount());
		}
	}

	private _adjustOffside(): void {
		this.startLineNumber++;
		for (const child of this.children) {
			child._adjustOffside();
		}
	}

	private _adjustRegular(document: AbstractDocument, maxEndLineNumber: number): void {
		if (this.endLineNumber < maxEndLineNumber) {
			const nextLine = document.getLineText(this.endLineNumber).trim();
			const isClosingBracket = /^[\}\]\)];?$/.test(nextLine);
			const isClosingTag = /^<\/\w/.test(nextLine);
			if (isClosingBracket || isClosingTag) {
				this.endLineNumber++;
			}
		}

		for (let i = this.children.length - 1; i >= 0; i--) {
			const child = this.children[i];
			const childMaxEndLineNumber = (i + 1 < this.children.length ? this.children[i + 1].startLineNumber - 1 : maxEndLineNumber);
			child._adjustRegular(document, childMaxEndLineNumber);
		}
	}

	toOverlayNode(document: AbstractDocument, isRoot: boolean): OverlayNode {
		const children: OverlayNode[] = [];
		let nextLineNumber = (isRoot && this.startLineNumber === 1 ? 1 : this.startLineNumber + 1);

		// for (let lineNumber = this.startLineNumber + 1;)
		for (const child of this.children) {

			// Generate a node for each skipped line
			for (let lineNumber = nextLineNumber; lineNumber < child.startLineNumber; lineNumber++) {
				const node = createOverlayNode(document, lineNumber, lineNumber, 'LINE', []);
				if (node) {
					children.push(node);
				}
			}

			children.push(child.toOverlayNode(document, false));

			nextLineNumber = child.endLineNumber + 1;
		}

		// Generate a node for each skipped line
		for (let lineNumber = nextLineNumber; lineNumber < this.endLineNumber; lineNumber++) {
			const node = createOverlayNode(document, lineNumber, lineNumber, 'LINE', []);
			if (node) {
				children.push(node);
			}
		}

		return createOverlayNode(document, this.startLineNumber, this.endLineNumber, 'FOLD', children);
	}
}

function createOverlayNode(doc: AbstractDocument, startLineNumber: number, endLineNumber: number, kind: string, children: OverlayNode[]): OverlayNode {
	const startOffset = doc.getOffsetAtPosition(new Position(startLineNumber - 1, 0));
	const endPosition = (
		endLineNumber < doc.getLineCount()
			? new Position(endLineNumber, 0)
			: new Position(endLineNumber - 1, doc.getLineLength(endLineNumber - 1))
	);
	const endOffset = doc.getOffsetAtPosition(endPosition);
	return new OverlayNode(
		startOffset,
		endOffset,
		kind,
		children
	);
}


export interface ISimpleTextModel {
	getLineCount(): number;
	getLineContent(lineNumber: number): string;
	getOptions(): { tabSize: number };
}

function generateFoldingRegions(model: ISimpleTextModel, languageId: string): FoldingRegions {
	return _computeRanges(model, isOffSide(languageId));
}

function isOffSide(languageId: string): boolean {
	return ['clojure', 'coffeescript', 'fsharp', 'latex', 'markdown', 'pug', 'python', 'sql', 'yaml'].includes(languageId);
}

function _computeRanges(model: ISimpleTextModel, offSide: boolean): FoldingRegions {
	const tabSize = model.getOptions().tabSize;
	const result = new RangesCollector();

	const previousRegions: PreviousRegion[] = [];
	const line = model.getLineCount() + 1;
	previousRegions.push({ indent: -1, endAbove: line, line }); // sentinel, to make sure there's at least one entry

	for (let line = model.getLineCount(); line > 0; line--) {
		const lineContent = model.getLineContent(line);
		const indent = computeIndentLevel(lineContent, tabSize);
		let previous = previousRegions[previousRegions.length - 1];
		if (indent === -1) {
			if (offSide) {
				// for offSide languages, empty lines are associated to the previous block
				// note: the next block is already written to the results, so this only
				// impacts the end position of the block before
				previous.endAbove = line;
			}
			continue; // only whitespace
		}
		if (previous.indent > indent) {
			// discard all regions with larger indent
			do {
				previousRegions.pop();
				previous = previousRegions[previousRegions.length - 1];
			} while (previous.indent > indent);

			// new folding range
			const endLineNumber = previous.endAbove - 1;
			if (endLineNumber - line >= 1) { // needs at east size 1
				result.insertFirst(line, endLineNumber, indent);
			}
		}
		if (previous.indent === indent) {
			previous.endAbove = line;
		} else { // previous.indent < indent
			// new region with a bigger indent
			previousRegions.push({ indent, endAbove: line, line });
		}
	}
	return result.toIndentRanges();
}

interface PreviousRegion {
	indent: number; // indent or -2 if a marker
	endAbove: number; // end line number for the region above
	line: number; // start line of the region. Only used for marker regions.
}

const MAX_FOLDING_REGIONS = 0xFFFF;
const MAX_LINE_NUMBER = 0xFFFFFF;
const MASK_INDENT = 0xFF000000;

class RangesCollector {
	private readonly _startIndexes: number[];
	private readonly _endIndexes: number[];
	private readonly _indentOccurrences: number[];
	private _length: number;

	constructor() {
		this._startIndexes = [];
		this._endIndexes = [];
		this._indentOccurrences = [];
		this._length = 0;
	}

	public insertFirst(startLineNumber: number, endLineNumber: number, indent: number) {
		if (startLineNumber > MAX_LINE_NUMBER || endLineNumber > MAX_LINE_NUMBER) {
			return;
		}
		const index = this._length;
		this._startIndexes[index] = startLineNumber;
		this._endIndexes[index] = endLineNumber;
		this._length++;
		if (indent < 1000) {
			this._indentOccurrences[indent] = (this._indentOccurrences[indent] || 0) + 1;
		}
	}

	public toIndentRanges() {
		// reverse and create arrays of the exact length
		const startIndexes = new Uint32Array(this._length);
		const endIndexes = new Uint32Array(this._length);
		for (let i = this._length - 1, k = 0; i >= 0; i--, k++) {
			startIndexes[k] = this._startIndexes[i];
			endIndexes[k] = this._endIndexes[i];
		}
		return new FoldingRegions(startIndexes, endIndexes);
	}
}

/**
 * Returns:
 *  - -1 => the line consists of whitespace
 *  - otherwise => the indent level is returned value
 */
function computeIndentLevel(line: string, tabSize: number): number {
	let indent = 0;
	let i = 0;
	const len = line.length;

	while (i < len) {
		const chCode = line.charCodeAt(i);
		if (chCode === CharCode.Space) {
			indent++;
		} else if (chCode === CharCode.Tab) {
			indent = indent - indent % tabSize + tabSize;
		} else {
			break;
		}
		i++;
	}

	if (i === len) {
		return -1; // line only consists of whitespace
	}

	return indent;
}

class FoldingRegions {
	private readonly _startIndexes: Uint32Array;
	private readonly _endIndexes: Uint32Array;

	private _parentsComputed: boolean;

	constructor(startIndexes: Uint32Array, endIndexes: Uint32Array) {
		this._startIndexes = startIndexes;
		this._endIndexes = endIndexes;
		this._parentsComputed = false;
		// this.ensureParentIndices();
	}

	private ensureParentIndices() {
		if (!this._parentsComputed) {
			this._parentsComputed = true;
			const parentIndexes: number[] = [];
			const isInsideLast = (startLineNumber: number, endLineNumber: number) => {
				const index = parentIndexes[parentIndexes.length - 1];
				return this.getStartLineNumber(index) <= startLineNumber && this.getEndLineNumber(index) >= endLineNumber;
			};
			for (let i = 0, len = this._startIndexes.length; i < len; i++) {
				const startLineNumber = this._startIndexes[i];
				const endLineNumber = this._endIndexes[i];
				if (startLineNumber > MAX_LINE_NUMBER || endLineNumber > MAX_LINE_NUMBER) {
					throw new Error('startLineNumber or endLineNumber must not exceed ' + MAX_LINE_NUMBER);
				}
				while (parentIndexes.length > 0 && !isInsideLast(startLineNumber, endLineNumber)) {
					parentIndexes.pop();
				}
				const parentIndex = parentIndexes.length > 0 ? parentIndexes[parentIndexes.length - 1] : -1;
				parentIndexes.push(i);
				this._startIndexes[i] = startLineNumber + ((parentIndex & 0xFF) << 24);
				this._endIndexes[i] = endLineNumber + ((parentIndex & 0xFF00) << 16);
			}
		}
	}

	public get length(): number {
		return this._startIndexes.length;
	}

	public getStartLineNumber(index: number): number {
		return this._startIndexes[index] & MAX_LINE_NUMBER;
	}

	public getEndLineNumber(index: number): number {
		return this._endIndexes[index] & MAX_LINE_NUMBER;
	}

	public getParentIndex(index: number) {
		this.ensureParentIndices();
		const parent = ((this._startIndexes[index] & MASK_INDENT) >>> 24) + ((this._endIndexes[index] & MASK_INDENT) >>> 16);
		if (parent === MAX_FOLDING_REGIONS) {
			return -1;
		}
		return parent;
	}

	public contains(index: number, line: number) {
		return this.getStartLineNumber(index) <= line && this.getEndLineNumber(index) >= line;
	}

	private findIndex(line: number) {
		let low = 0, high = this._startIndexes.length;
		if (high === 0) {
			return -1; // no children
		}
		while (low < high) {
			const mid = Math.floor((low + high) / 2);
			if (line < this.getStartLineNumber(mid)) {
				high = mid;
			} else {
				low = mid + 1;
			}
		}
		return low - 1;
	}

	public findRange(line: number): number {
		let index = this.findIndex(line);
		if (index >= 0) {
			const endLineNumber = this.getEndLineNumber(index);
			if (endLineNumber >= line) {
				return index;
			}
			index = this.getParentIndex(index);
			while (index !== -1) {
				if (this.contains(index, line)) {
					return index;
				}
				index = this.getParentIndex(index);
			}
		}
		return -1;
	}
}
