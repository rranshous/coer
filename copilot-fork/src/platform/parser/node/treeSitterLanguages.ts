/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/

/**
 * Languages we can parse using tree-sitter. Each enum member corresponds to a tree-sitter parser.
 */
export enum WASMLanguage {
	Python = 'python',
	JavaScript = 'javascript', // Also includes jsx support
	TypeScript = 'typescript',
	TypeScriptTsx = 'tsx',
	Go = 'go',
	Ruby = 'ruby',
	Csharp = 'csharp',
	Cpp = 'cpp',
	Java = 'java',
	Rust = 'rust',
}

export class TreeSitterUnknownLanguageError extends Error {
	constructor(language: string) {
		super(`Unrecognized language: ${language}`);
	}
}

const languageIdToWasmLanguageMapping: { [language: string]: WASMLanguage } = {
	python: WASMLanguage.Python,
	javascript: WASMLanguage.JavaScript,
	javascriptreact: WASMLanguage.JavaScript,
	jsx: WASMLanguage.JavaScript,
	typescript: WASMLanguage.TypeScript,
	typescriptreact: WASMLanguage.TypeScriptTsx,
	tsx: WASMLanguage.TypeScriptTsx,
	go: WASMLanguage.Go,
	ruby: WASMLanguage.Ruby,
	csharp: WASMLanguage.Csharp,
	cpp: WASMLanguage.Cpp,
	java: WASMLanguage.Java,
	rust: WASMLanguage.Rust,
};

/**
 * @returns a {@link WASMLanguage} if can convert the language ID (from VS Code); otherwise, returns `undefined`.
 */
export function getWasmLanguage(languageId: string): WASMLanguage | undefined {
	if (languageId in languageIdToWasmLanguageMapping) {
		return languageIdToWasmLanguageMapping[languageId];
	}
}
