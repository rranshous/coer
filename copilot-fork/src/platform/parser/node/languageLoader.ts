/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/

import * as path from '../../../util/vs/base/common/path';
import { WASMLanguage } from './treeSitterLanguages';
import Parser = require('web-tree-sitter');

export class LanguageLoader {

	private readonly loadedLanguagesCache;

	constructor() {
		this.loadedLanguagesCache = new Map<WASMLanguage, Promise<Parser.Language>>();
	}

	loadLanguage(wasmLanguage: WASMLanguage): Promise<Parser.Language> {
		if (!this.loadedLanguagesCache.has(wasmLanguage)) {
			this.loadedLanguagesCache.set(wasmLanguage, this._doLoadLanguage(wasmLanguage));
		}
		return this.loadedLanguagesCache.get(wasmLanguage)!;
	}

	private _doLoadLanguage(language: WASMLanguage): Promise<Parser.Language> {
		// construct a path that works both for the TypeScript source, which lives under `/src`, and for
		// the transpiled JavaScript, which lives under `/dist`
		const wasmFileLang = language === 'csharp' ? 'c-sharp' : language;

		const wasmFilename = `tree-sitter-${wasmFileLang}.wasm`;

		// depending on if file is being run from the webpack bundle or source, change the relative path
		const wasmFile =
			path.basename(__dirname) === 'dist'
				? path.resolve(__dirname, wasmFilename)
				: path.resolve(__dirname, '../../../../dist', wasmFilename);

		return Parser.Language.load(wasmFile);
	}
}
