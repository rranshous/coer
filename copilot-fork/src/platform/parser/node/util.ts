/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/

import type { SyntaxNode } from 'web-tree-sitter';
import { WASMLanguage } from './treeSitterLanguages';

/**
 * Extracts the identifier for the given node (is based on heuristics, so not 100% accurate for all languages)
 */
export function extractIdentifier(node: SyntaxNode, languageId: string): string | undefined {
	switch (languageId) {
		case 'python':
		case 'csharp':
			return node.children.find(c => c.type.match(/identifier/))?.text;
		case 'go': {
			const identifierChild = node.children.find(c => c.type.match(/identifier/));
			if (identifierChild) { return identifierChild.text; }
			const specChild = node.children.find(c => c.type.match(/spec/));
			return specChild?.children.find(c => c.type.match(/identifier/))?.text;
		}
		case 'javascript':
		case 'javascriptreact':
		case 'typescript':
		case 'typescriptreact':
		case 'cpp': {
			const declarator = node.children.find(c => c.type.match(/declarator/));
			if (declarator) { return declarator.children.find(c => c.type.match(/identifier/))?.text; }
			const identifierChild = node.children.find(c => c.type.match(/identifier/));
			return identifierChild?.text;
		}
		case 'java': {
			/*
				handles
				```java
				// class identifier
				class Fo<<>>o { }

				class Foo {
					// method identifier
					public void ba<<>>r() { }
				}

				// enum identifier
				enum F<<>>oo {
					// enum constant identifier
					BA<<>>Z
				}

				// interface identifier
				interface Fo<<>>o { }
				```
			*/
			const identifierChild = node.children.find(c => c.type === 'identifier');
			return identifierChild?.text;
		}
		case 'ruby':
			return node.children.find(c => c.type.match(/constant|identifier/))?.text;
		default:
			return node.children.find(c => c.type.match(/identifier/))?.text;
	}
}


export function isDocumentableNode(node: SyntaxNode, language: WASMLanguage) {
	switch (language) {
		case WASMLanguage.TypeScript:
		case WASMLanguage.TypeScriptTsx:
		case WASMLanguage.JavaScript:
			return node.type.match(/definition|declaration|declarator|export_statement/);
		case WASMLanguage.Go:
			return node.type.match(/definition|declaration|declarator|var_spec/);
		case WASMLanguage.Cpp:
			return node.type.match(/definition|declaration|class_specifier/);
		case WASMLanguage.Ruby:
			return node.type.match(/module|class|method|assignment/);
		default:
			return node.type.match(/definition|declaration|declarator/);
	}
}
