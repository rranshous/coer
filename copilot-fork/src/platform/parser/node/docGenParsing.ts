/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/

import type { SyntaxNode } from 'web-tree-sitter';
import { Node, TreeSitterOffsetRange } from './nodes';
import { _parse } from './parserWithCaching';
import { _getNodeMatchingSelection } from './selectionParsing';
import { WASMLanguage } from './treeSitterLanguages';
import { extractIdentifier, isDocumentableNode } from './util';


export type NodeToDocumentContext = {

	/** is undefined when we couldn't determine the identifier */
	nodeIdentifier: string | undefined;

	nodeToDocument: Node;

	/**
	 * 'expanding' - selection was expanded to a wrapping node
	 * 'matchingSelection' - node was picked by observing range overlap with the selection
	 */
	nodeSelectionBy: 'expanding' | 'matchingSelection';
};

/**
 * Starting from the smallest AST node that wraps `selection` and climbs up the AST until it sees a "documentable" node.
 * See {@link isDocumentableNode} for definition of a "documentable" node.
 *
 * @param language The language ID of the source code.
 * @param source The source code to parse.
 * @param selection The range to document.
 * @returns An object containing the smallest node containing the selection range, its parent node, the node to document, and the number of nodes climbed up to reach the documentable node.
 * 			Returns undefined if the language ID is not supported.
 */
export async function _getNodeToDocument(
	language: WASMLanguage,
	source: string,
	selection: TreeSitterOffsetRange
): Promise<NodeToDocumentContext> {

	const treeRef = await _parse(language, source);

	try {

		// if selection is non-empty, try identify a documentable AST node that most matches the selection
		// otherwise, try to find the smallest documentable AST node that wraps the selection

		const isSelectionEmpty = selection.startIndex === selection.endIndex;

		const selectionMatchedNode = isSelectionEmpty ? undefined : _getNodeMatchingSelection(treeRef.tree, selection, language);

		if (selectionMatchedNode) {
			const nodeIdentifier = extractIdentifier(selectionMatchedNode, language);
			return {
				nodeIdentifier,
				nodeToDocument: Node.ofSyntaxNode(selectionMatchedNode),
				nodeSelectionBy: 'matchingSelection'
			};
		}

		const nodeContainingCursor = treeRef.tree.rootNode.descendantForIndex(selection.startIndex, selection.endIndex);

		let nodeToDocument: SyntaxNode = nodeContainingCursor;
		let nNodesClimbedUp = 0;

		// ascend the parse tree until we find a declaration/definition (documentable) node or reach the root node
		while (!isDocumentableNode(nodeToDocument, language) && nodeToDocument.parent !== null) {
			nodeToDocument = nodeToDocument.parent;
			++nNodesClimbedUp;
		}

		const nodeIdentifier = extractIdentifier(nodeToDocument, language);
		return {
			nodeIdentifier,
			nodeToDocument: Node.ofSyntaxNode(nodeToDocument),
			nodeSelectionBy: 'expanding',
		};
	} finally {
		treeRef.dispose();
	}
}

export async function _getDocumentableNodeIfOnIdentifier(
	language: WASMLanguage,
	source: string,
	range: TreeSitterOffsetRange
): Promise<{ identifier: string; nodeRange?: TreeSitterOffsetRange } | undefined> {
	const treeRef = await _parse(language, source);
	try {

		const smallestNodeContainingRange = treeRef.tree.rootNode.descendantForIndex(range.startIndex, range.endIndex);

		if (smallestNodeContainingRange.type.match(/identifier/) &&
			(smallestNodeContainingRange.parent === null || isDocumentableNode(smallestNodeContainingRange.parent, language))
		) {
			const parent = smallestNodeContainingRange.parent;

			const parentNodeRange = parent === null
				? undefined
				: { startIndex: parent.startIndex, endIndex: parent.endIndex };

			return {
				identifier: smallestNodeContainingRange.text,
				nodeRange: parentNodeRange
			};
		}
	} finally {
		treeRef.dispose();
	}
}
