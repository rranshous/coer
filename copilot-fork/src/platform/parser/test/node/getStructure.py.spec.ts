/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/

import { afterAll, describe, expect, test } from 'vitest';
import { _dispose } from '../../node/parserImpl';
import { WASMLanguage } from '../../node/treeSitterLanguages';
import { fromFixture, snapshotPathInFixture, srcWithAnnotatedStructure } from './getStructure.util';

describe('getStructure - python', () => {

	afterAll(() => _dispose());

	function pySrcWithStructure(source: string) {
		return srcWithAnnotatedStructure(WASMLanguage.Python, source);
	}

	test('py source with different syntax constructs', async () => {

		const source = await fromFixture('test.py');

		expect(await pySrcWithStructure(source)).toMatchSnapshot();
	});

	test('try-catch block', async () => {

		const file = 'try.py';

		const source = await fromFixture(file);

		await expect(await pySrcWithStructure(source)).toMatchFileSnapshot(snapshotPathInFixture(file));
	});
});
