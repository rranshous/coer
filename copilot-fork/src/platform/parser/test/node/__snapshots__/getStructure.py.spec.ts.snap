// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getStructure - python > py source with different syntax constructs 1`] = `
"<COMMENT># This is a single line comment
</COMMENT><EXPRESSION_STATEMENT>
"""
This is a multi-line comment
"""
</EXPRESSION_STATEMENT><COMMENT-1>
# Importing a module
</COMMENT-1><IMPORT_STATEMENT>import math
</IMPORT_STATEMENT><COMMENT-2>
# Variable assignment
</COMMENT-2><EXPRESSION_STATEMENT-1>x = 10
</EXPRESSION_STATEMENT-1><COMMENT-3>
# Function definition
</COMMENT-3><FUNCTION_DEFINITION>def square(num):
    <EXPRESSION_STATEMENT-2>"""This is a docstring"""
</EXPRESSION_STATEMENT-2><EXPRESSION_STATEMENT-3>    foo = a or b
</EXPRESSION_STATEMENT-3><RETURN_STATEMENT>    return num ** 2</RETURN_STATEMENT>
</FUNCTION_DEFINITION><COMMENT-4>
# Conditional statements
</COMMENT-4><IF_STATEMENT>if x > 0:
    <EXPRESSION_STATEMENT-4>print("Positive")</EXPRESSION_STATEMENT-4>
elif x < 0:
    <EXPRESSION_STATEMENT-5>print("Negative")</EXPRESSION_STATEMENT-5>
else:
    <EXPRESSION_STATEMENT-6>print("Zero")</EXPRESSION_STATEMENT-6>
</IF_STATEMENT><COMMENT-5>
# For loop
</COMMENT-5><FOR_STATEMENT>for i in range(5):
    <EXPRESSION_STATEMENT-7>print(i)</EXPRESSION_STATEMENT-7>
</FOR_STATEMENT><COMMENT-6>
# While loop
</COMMENT-6><WHILE_STATEMENT>while x > 0:
    <EXPRESSION_STATEMENT-8>x -= 1</EXPRESSION_STATEMENT-8>
</WHILE_STATEMENT><COMMENT-7>
# List comprehension
</COMMENT-7><EXPRESSION_STATEMENT-9>squares = [i**2 for i in range(10)]
</EXPRESSION_STATEMENT-9><COMMENT-8>
# Class definition
</COMMENT-8><CLASS_DEFINITION>class MyClass:
    <FUNCTION_DEFINITION-1>def __init__(self, name):
        <EXPRESSION_STATEMENT-10>self.name = name</EXPRESSION_STATEMENT-10>
</FUNCTION_DEFINITION-1><FUNCTION_DEFINITION-2>
    def greet(self):
        <EXPRESSION_STATEMENT-11>print(f"Hello, {self.name}")</EXPRESSION_STATEMENT-11></FUNCTION_DEFINITION-2>
</CLASS_DEFINITION><COMMENT-9>
# Creating an object
</COMMENT-9><EXPRESSION_STATEMENT-12>obj = MyClass("Python")
</EXPRESSION_STATEMENT-12><COMMENT-10>
# Calling a method
</COMMENT-10><EXPRESSION_STATEMENT-13>obj.greet()
</EXPRESSION_STATEMENT-13><COMMENT-11>
# Error handling
</COMMENT-11><TRY_STATEMENT>try:
    <EXPRESSION_STATEMENT-14>print(10 / 0)</EXPRESSION_STATEMENT-14>
except ZeroDivisionError:
    <EXPRESSION_STATEMENT-15>print("Cannot divide by zero")</EXPRESSION_STATEMENT-15>
finally:
    <EXPRESSION_STATEMENT-16>print("End of error handling")</EXPRESSION_STATEMENT-16>
</TRY_STATEMENT><COMMENT-12>
# With statement
</COMMENT-12><WITH_STATEMENT>with open('file.txt', 'w') as f:
    <EXPRESSION_STATEMENT-17>f.write("Hello, World!")</EXPRESSION_STATEMENT-17>
</WITH_STATEMENT><COMMENT-13>
# Lambda function
</COMMENT-13><EXPRESSION_STATEMENT-18>square = lambda x: x**2
</EXPRESSION_STATEMENT-18><COMMENT-14>
# Generator expression
</COMMENT-14><EXPRESSION_STATEMENT-19>gen = (i**2 for i in range(10))
</EXPRESSION_STATEMENT-19><COMMENT-15>
# Decorator
</COMMENT-15><FUNCTION_DEFINITION-3>def my_decorator(func):
    <FUNCTION_DEFINITION-4>def wrapper():
        <EXPRESSION_STATEMENT-20>print("Before function call")
</EXPRESSION_STATEMENT-20><EXPRESSION_STATEMENT-21>        func()
</EXPRESSION_STATEMENT-21><EXPRESSION_STATEMENT-22>        print("After function call")</EXPRESSION_STATEMENT-22>
</FUNCTION_DEFINITION-4><RETURN_STATEMENT-1>    return wrapper</RETURN_STATEMENT-1>
</FUNCTION_DEFINITION-3><DECORATED_DEFINITION>
@my_decorator
<FUNCTION_DEFINITION-5>def say_hello():
    <EXPRESSION_STATEMENT-23>print("Hello")</EXPRESSION_STATEMENT-23></FUNCTION_DEFINITION-5>
</DECORATED_DEFINITION><EXPRESSION_STATEMENT-24>
say_hello()</EXPRESSION_STATEMENT-24>
"
`;
